### PR TITLE
grc: Call connect() for hier blocks that don't have ports

### PR DESCRIPTION
**Pull request by haakov**
Fixes https://github.com/gnuradio/gnuradio/issues/2950.

### CODE REVIEW

**Comment by skoslowski** [comment]
Maybe these kind of connect calls should be added as a template to the stream domain spec.

**Comment by haakov** [comment]
I thought about that. But wouldn't that cause problems for blocks that only have message ports?

**Comment by marcusmueller** [comment]
I see the message-port-only problem; but what about e.g. RFNoC blocks?

**Comment by haakov** [comment]
I'm afraid I haven't used RFNoC blocks yet. Could you explain how they relate to this?

**Comment by bastibl** [comment]
There are also GRC Blocks that are just there to inject code in the flowgraph template (e.g. CtrlPort Monitor). They might also have code in `make`. Not sure if it would work with this logic.
https://github.com/gnuradio/gnuradio/blob/master/gr-blocks/grc/blocks_ctrlport_performance.block.yml

**Comment by nickoe** [comment]
@haakov Ping

**Comment by haakov** [comment]
@nickoe thanks for the reminder!

Maybe it is easier to just add the `connect()` call in the hier block's make template.
Like this: https://github.com/haakov/gnuradio/commit/3718527ceb54b8e4be2bbe43875b84eb8288a060 (C++ part not tested yet)

If there are other blocks that need to be connected, the call can be added in their YAML files too, but I can't think of any right now.

**Comment by haakov** [comment]
Updated to only affect hier blocks.

**Comment by haakov** [comment]
oop, just realized this won't work for hier blocks with exclusively optional ports. Never mind

**Comment by mbr0wn** [comment]
> oop, just realized this won't work for hier blocks with exclusively optional ports. Never mind

Looks like this will be more work then?

**Comment by haakov** [comment]
Actually, no. I'd like to withdraw my previous comment :)

Turns out that connecting a hier block with exclusively optional ports is not allowed anyways:
```
  File "/usr/local/lib/python3.9/site-packages/gnuradio/gr/hier_block2.py", line 100, in connect
    self.primitive_connect(*args)
ValueError: Block Sink(1) must not have any input or output ports
```
This PR is ready for review.

File: grc/core/generator/cpp_hier_block.py [hunk 1]
```diff
@@ -124,6 +124,9 @@ def var_or_value(name):
             )
         else:
             t['make'] = '{cls}()'.format(cls=block_id)
+        # Self-connect if there aren't any ports
+        if {len(data['inputs']), len(data['outputs'])} == {0}:
```
Comment (skoslowski) [inline grc/core/generator/cpp_hier_block.py:128]: IMO this does not convey your intend very well. May I suggest (for readability) to use  
`if not data["inputs"] and not data["outputs"]:`

**Review by skoslowski — CHANGES_REQUESTED** [review]
I can confirm, that this adds the desired single-arg connect() call. I used a hier_block with a message strobe and message debug, which I put in a otherwise empty top_block. Only with this patch applied output is generated.

File: grc/core/generator/cpp_hier_block.py (hunk 1 above)
Comment (haakov) [reply grc/core/generator/cpp_hier_block.py:128]: Agreed! I'll fix it

Comment (mbr0wn) [reply grc/core/generator/cpp_hier_block.py:128]: When you do, can you please rebase onto the latest master?

**Pull request merged**